### DISCUSSION TITLE
Zoom

### DISCUSSION THREAD

**Discussion by HGSchuetz** [sumatrapdfreader/sumatrapdf#4011] [Ideas]
Hi,

using Zoom, either with keyboard or pinch on touchscreen, starts from top / left or left position.

For keyboard its ok, but for pinch on touchscreen it would be very helpful, if it zooms ouf of the position of the fingers.

Im zooming large pdf-maps. So, it is not very helpful when you want to zoom in the very right side, but Sumatra starts zooming on the left side. So I need to swipe over all to the right again when zoomed in.

When double touch on screen it selects the image of the pdf-map. A setting which allows the double click to go back to FitContent (as per settings) instead, would be a perfect extension!

Regards
Hans

**Comment by HGSchuetz** [comment]
Hi,

any chance implementing the "zoom at finger pos" feature?

BR Hans

**Reply by kjk** [reply]
It seems that zoom with pinch should use mouse position as zoom center, like Chrome.

If that's acceptable, please open an issue via https://github.com/sumatrapdfreader/sumatrapdf/issues (that's not a promise I can implement it any time soo)

**Reply by GitHubRulesOK** [reply]
there should be an old issue already open ?
see
https://github.com/sumatrapdfreader/sumatrapdf/issues/481
 https://github.com/sumatrapdfreader/sumatrapdf/issues/861
related https://github.com/sumatrapdfreader/sumatrapdf/issues/3983

**Comment by kjk** [comment]
I've changed pinch zoom to zoom around the pinch position. I'm not quite happy how pinch works but that's not related to picking the point around which to zoom.

Note it already worked like that for touchpad zoom and it's more accurate because it zooms around mouse position.

Also note that keyboard zoom will zoom around selection (i.e. zoom around center of selection).

I'm considering a new `ZoomInAroundMouse` / `ZoomOutAroundMouse` that would zoom around mouse position and would allow you to bind them to keyboard shortcuts.

Or just change the behavior of `ZoomIn` / `ZoomOut` commands. On one hand I don't want to change how things work because some people might be used to it. On the other hand this looks like a better behavior and a bit obscure anyway.

Any idea how other PDF readers behave?

**Comment by kjk** [comment]
Ok, I fixed pinch zoom so I'm happy how it works now. Should be available in next [pre-release](https://www.sumatrapdfreader.org/prerelease) 16068 or later.

**Comment by HGSchuetz** [comment]
Woow, that sounds excellent!

Thanks for your response and implementing this feature! I like Sumatra very much, and this would for sure be another vital feature for all tablet users out there!

BR Hans

**Comment by kjk** [comment]
Some final zoom improvements. 

Latest behavior (in 16073 or later) should be:
* touchscreen zooms around touched point, zooms smoothly
* zoom with touchpad zooms around cursor position, zooms smoothly (new in pre-release)
* zooming with keyboard:
  * zooms by steps
  * if there's selection, we zoom around center of selection
  * if no selection, we zoom around mouse position (new in pre-release) i.e. we try to maintain the point on page under mouse in the same position on screen)
    * if mouse not over a page, zoom around center (was: top-left of first visible page)

"Zooms by steps" means that we zoom to the next zoom level (as shown in menu).

"Zooms smoothly" means we zoom by scale from pinch touchscreen or touchbar gesture.
